### DISCUSSION TITLE
Hierarchical tags

### DISCUSSION THREAD

**Discussion by thiloplanz** [jotaen/klog#55] [Feature ideas & discussions]
It would be nice to have support for hierarchical tags.

As a first step, not necessarily with full support in the tool itself, but just the ability to have tags like `#cs/cs123` or `#cs:cs123`.

Currently, this is not parsed as a tag at all (or maybe just as `#cs`).

So when I record time doing customer support (`#cs`) on support ticket 123 (`#cs123`), I have to remember to add both tags (or use some shell scripting to do the report aggregation in a post-processing step).

An example where something similar works very well is how you can nest accounts in https://www.ledger-cli.org/

**Comment by jotaen** [comment]
Some background info, just for curiosity: The tags being non-hierarchical was done on purpose. I had considered a hierarchical structure initially (such as the accounts in ledger), but thought it would be better to give users most flexibility and allow them to come up with their own system. Note, that even if there was something like `#cs:123` you could accidentally forget the `cs:` prefix. Unlike with accounts in ledger you cannot register tags upfront in klog and thus enforce a certain structure. This in my opinion makes sense, because tags are completely optional in klog, but accounts in ledger are mandatory.

As you already pointed out, tags can only contain [letters, digits and underscores](https://github.com/jotaen/klog/blob/main/Specification.md#tags). This is important so that you can embed tags naturally in the flow of sentences, e.g.:

```
2020-01-01
    2h Went to the #gym: #bizeps, #trizeps, #shoulders.
```

That being said: what I could imagine to be a good “middle ground” would be to support wildcards in the command line tool for querying. That way you could use an underscore as delimiter `#cs_123` and query for it like so:

```
klog total --tag 'cs_*' work.klg
```

Thoughts?

**Reply by thiloplanz** [reply]
I had not thought to try `_`. That will work for me for the moment. 

> This is important so that you can embed tags naturally in the flow of sentences, e.g.:

yeah, different use case. I have zero prose in these files, it's lot of small entries tagged by different dimensions (activity type, bug tracking number, project name, customer name, etc) that I never look at again (except in aggregated form in the various reports later).

**Reply by vladdeSV** [reply]
Looking at the comment https://github.com/jotaen/klog/discussions/54#discussioncomment-570317, where you can filter for multiple tags, seems like it does this already, only with a slightly different syntax.

Instead of
```klog
2021-04-05
  1h #cs:123
```
```sh
klog total --tag 'cs:123' work.klg
```

you do 

```klog
2021-04-05
  1h #cs #123
```
```sh
klog total --tag 'cs' --tag '123' work.klg
```

**Reply by thiloplanz** [reply]
@vladdeSV Yes, one can filter by multiple tags, but the main point of this feature request was to have a hierarchy within tags, so that it is not necessary to have multiple tags in the first place.

```
2021-04-05
  1h #cs:123
  1h #cs:234
```

`klog total --tag 'cs'` would now return the sum across all `cs:*` tags.

The proposed wildcard syntax on the reporting side would also facilitate that.

**Reply by jotaen** [reply]
I think having wildcard filters would make a nice enhancement, thanks for the idea: https://github.com/jotaen/klog/issues/56
